eval: fix evaluation of ALL, ANY, and SOME with NULLs on the left

yuzefovich opened the pull request:
**parser: fix parsing ~ when used as comparison operator in some cases**

This commit fixes how we parse expressions like
```
NOT ('a' ~ ANY (ARRAY['x']:::STRING[]))
```
which is equivalent to
```
('a' !~ ANY (ARRAY['x']:::STRING[]))
```
which we parse correctly. The problem was that we parse `~` symbol as
"unary complement" math operator and not as "reg match" comparison
operator, and this is now fixed in a somewhat hacky way. Namely, we
still parse it as unary complement, but then when we're in the comparison
operator with sub-operator context, we retype it correctly. This seems
acceptable given that a more "proper" fix would probably require
non-trivial changes to the grammar.

Release note: None

**eval: fix evaluation of ALL, ANY, and SOME with NULL on the left**

Previously, we could run into an internal error (or a crash when the
vectorized engine is not used) when evaluating ALL, ANY, or SOME
comparison operators with some sub-operators (e.g. RegMatch). We forgot
to check whether the left argument is NULL, and this is now fixed. For
other sub-operators we seem to have done the right thing, so this bug
seems rather rare in practice, thus, I omitted the release note.

Fixes: #95158.

Release note: None

yuzefovich committed:
eval: fix evaluation of ALL, ANY, and SOME with NULL on the left

Previously, we could run into an internal error (or a crash when the
vectorized engine is not used) when evaluating ALL, ANY, or SOME
comparison operators with some sub-operators (e.g. RegMatch). We forgot
to check whether the left argument is NULL, and this is now fixed. For
other sub-operators we seem to have done the right thing, so this bug
seems rather rare in practice, thus, I omitted the release note.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95180)
<!-- Reviewable:end -->

rafiss commented:
this seems related to https://github.com/cockroachdb/cockroach/issues/37793 (my first CRDB bug report 🙂)

yuzefovich committed:
parser: fix parsing `~` when used as comparison operator in some cases

This commit fixes how we parse expressions like
```
NOT ('a' ~ ANY (ARRAY['x']:::STRING[]))
```
which is equivalent to
```
('a' !~ ANY (ARRAY['x']:::STRING[]))
```
which we parse correctly. The problem was that we parse `~` symbol as
"unary complement" math operator and not as "reg match" comparison
operator, and this is now fixed in a somewhat hacky way. Namely, we
still parse it as unary complement, but then when we're in the comparison
operator with sub-operator context, we retype it correctly. This seems
acceptable given that a more "proper" fix would probably require
non-trivial changes to the grammar.

Release note: None

yuzefovich commented:
I ended up taking out the commit that would fix #37793 since my WIP change introduced some regressions I couldn't figure out. Also, the tests added in the (now second) commit exposed a parser bug when queries executed in a distributed fashion, so I added another (now first) commit to fix that.

rharding6373 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95180#-:-NMaJdWUBlcEOAEwy82W:b-kbynj8)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+

mgartner commented:
Interesting. I spent some time trying to figure out how Postgres's grammar handles this case, but didn't get far. I don't see explicit support for the `~` unary operator in their grammar. https://github.com/postgres/postgres/blob/15eb1228800a35042ef318f941173aa8b89002d1/src/backend/parser/gram.y#L14422-L14425

But it clearly is supported:

```
marcus=# select ~10;
 ?column?
----------
      -11
(1 row
```

yuzefovich commented:
I think the second commit in this PR fixes #97011, so I'm going to backport this to 22.2.

yuzefovich commented:
blathers backport 22.2
